experimental/dds/tree/src/id-compressor/AppendOnlySortedMap.ts

```diff
@@ -125,7 +139,7 @@ export class AppendOnlySortedMap<K, V> {
 	public append(key: K, value: V): void {
 		const { elements } = this;
 		const { length } = elements;
-		if (length !== 0 && this.comparator(key, elements[length - 2] as K) <= 0) {
+		if (length !== 0 && this.comparator(key, this.maxKey() as K) <= 0) {
```

Comment:
Our usual approach would be `this.maxKey ?? fail()` rather than `this.maxKey() as K`. Now that I'm actually thinking about it, the cast looks cleaner than the fail, and is cheaper. But it won't fail-fast in case that cast ever becomes incorrect for some reason. 

It seems to me that if we prefer the `as K` over `?? fail()` (or `assertNotUndefined(...)`, etc.), then what we actually want is `!this.maxKey()`, which is both syntactically cleaner and a more precise type conversion. But I think our linter disallows that. So... I say we be honest with ourselves and either disable that lint rule, or use `?? fail()`

Reply:
likewise for similar patterns in this file

Reply:
Personally, I have `?? fail()` as it's usually found in places the dev is _sure_ that it can't be null and is just fighting the type system. I like the cast better, so I'll leave it for now and we can use your other suggestion in a future PR.

Reply:
Yea, exactly, so we should just be using `!` then. We should bring that up with the team at standup to see if anyone has any problems with removing that lint rule.